docs/get_started/stable_c_abi.rst

```diff
@@ -0,0 +1,189 @@
+- Next 32 bits: null padding, or reserved for special flags in rare cases.
+- Last 64 bits: payload that is either a 64-bit integer, a 64-bit floating-point number, or a pointer to a heap-allocated object.
+
+.. figure:: ../_static/tvm-ffi-layout-any.svg
+   :alt: Layout of the 128-bit Any tagged union
+   :name: fig:layout-any
+
+   Figure 1. Layout spec for the ``Any`` type.
+
+The following conventions apply when representing values in ``Any``:
+
+- Primitive types: the last 64 bits directly store the value, for example:
+
+  * Integers
+  * Floating-point numbers
+
+- Heap-allocated objects: the last 64 bits store a pointer to the actual object, for example:
+
+  * `DLPack tensors <https://data-apis.org/array-api/2024.12/design_topics/data_interchange.html#dlpack-an-in-memory-tensor-structure>`_
+  * C-style strings
+
+- Arbitrary objects: the type index identifies the concrete type, and the last 64 bits store a pointer to a reference-counted object in TVM-FFI's object format, for example:
+
+  * :py:class:`tvm_ffi.Function`, representing all functions, such as Python/C++ functions/lambdas, etc.;
+  * :py:class:`tvm_ffi.Array` and :py:class:`tvm_ffi.Map` as list/dictionary containers of all ``Any`` values;
+  * Up to 2^32 types can be represented in this type system.
+
+Function Calling Convention
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Function calls in TVM-FFI share the same calling convention, ``universal_c_abi``, as described above.
+
+- ``handle: void*``: a library handle that contains the function to be called. It is usually ``nullptr`` for most use cases, but some libraries (such as cuDNN) may use it to store global contexts.
+- ``args: Any*``: a pointer to an array of ``Any`` values for input arguments, stored contiguously on stack or heap.
+- ``num_args: int``: number of input arguments.
+- ``result: Any*``: a pointer to a single ``Any`` that receives the output result.
+
+.. figure:: ../_static/tvm-ffi-layout-func.svg
+   :alt: Calling convention for universal_c_abi
+   :name: fig:layout-func
+
+   Figure 2. Layout spec of ``universal_c_abi`` function.
+
+
+Stability and Interoperability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Stability.** The pure C layout and the calling convention are stable across compiler versions and independent of host languages or frameworks.
+
+**Cross-language.** TVM-FFI implements this calling convention in multiple languages (C, C++, Python, Rust, ...), enabling code written in one language—or generated by a DSL targeting the ABI—to be called from another language.
+
+**Cross-framework.** TVM-FFI uses standard data structures such as `DLPack tensors <https://data-apis.org/array-api/2024.12/design_topics/data_interchange.html#dlpack-an-in-memory-tensor-structure>`_ to represent arrays, so compiled functions can be used from any array framework that implements the DLPack protocol (NumPy, PyTorch, TensorFlow, CuPy, JAX, and others).
+
+
+Stable ABI in C Code
+--------------------
+
```

Comment:
We design the C ABI with DSL and ML compilers in mind. In the context of ML compilers, usually we do not have access to c++ templates and needs to operate directly in LLVM (MLIR dialect) or low-level C. In such cases, it is helpful to directly leverage and target the stable C ABI.

This section shows how we can expose a kernel directly using the C ABI. Because we write code directly in C without any boilerplate, the code can serve as a direct reference for DSL compilers to expose kernel functions following our C ABI standard.